external-crates/move/move-vm/runtime/src/interpreter.rs

```diff
@@ -97,11 +99,128 @@ impl<'a, 'b> TypeView for TypeWithLoader<'a, 'b> {
     }
 }
 
+pub trait InterpreterInterface {
+    fn get_values_len(&self) -> usize;
+    fn set_location(&self, err: PartialVMError) -> VMError;
+    fn maybe_core_dump(&self, err: VMError, current_frame: &Frame) -> VMError;
+    fn get_internal_state(&self) -> ExecutionState;
+}
+
+impl InterpreterInterface for Interpreter {
+    fn get_values_len(&self) -> usize {
+        self.get_values_len()
+    }
+
+    fn set_location(&self, err: PartialVMError) -> VMError {
+        self.set_location(err)
+    }
+
+    fn maybe_core_dump(&self, err: VMError, current_frame: &Frame) -> VMError {
+        self.maybe_core_dump(err, current_frame)
+    }
+
+    fn get_internal_state(&self) -> ExecutionState {
+        self.get_internal_state()
+    }
+}
+
 impl Interpreter {
     /// Limits imposed at runtime
     pub fn runtime_limits_config(&self) -> &VMRuntimeLimitsConfig {
         &self.runtime_limits_config
     }
+
+    pub fn pre_hook_entrypoint(
+        _interpreter: &mut Interpreter,
+        plugins: &mut Vec<Box<dyn Plugin>>,
+        gas_meter: &mut impl GasMeter,
+        function: &Arc<Function>,
+        ty_args: &[Type],
+        data_store: &mut impl DataStore,
+        loader: &Loader,
+    ) -> VMResult<()> {
+        profile_open_frame!(gas_meter, function.pretty_string());
+
+        for plugin in plugins.iter_mut() {
+            plugin.pre_hook_entrypoint(function, ty_args, data_store.link_context(), loader)?;
```

Comment:
@wlmyng from our discussion, I have some concerns about handling failures because we probably don't always want to crash the entire interpreter because a plugin failed.
For example a profiler failing vs gas charging failing should be different severity levels.
Considered my suggestion of having different failure levels which each plugin returns, and the driver can decide how to handle?

Reply:
Yes for sure. I had tabled the error handling to get this draft out, but will circle back for them. Agree that different plugins beget different error severities